lib/Doctrine/Common/Cache/FileCache.php

```diff
@@ -126,9 +113,9 @@ protected function getFilename($id)
     {
         return $this->directory
             . DIRECTORY_SEPARATOR
-            . implode(str_split(hash('sha256', $id), 2), DIRECTORY_SEPARATOR)
+            . substr(hash('sha256', $id), 0, 2)
             . DIRECTORY_SEPARATOR
-            . preg_replace($this->disallowedCharacterPatterns, $this->replacementCharacters, $id)
+            . rawurlencode($id)
```

Comment:
With this we ensure 
- that the filename is already UNIQUE based on the cache key
- only valid characters inside the filename (no NULL byte, no `/` and also not the other chars that are disallowed on Windows)
- the filename still resembles the cache key, i.e. valid chars stay as-is (good for debugging)
- faster than a `preg_replace`
- the `preg_replace` implementation was actually useless as it didn't prevent collisions at all, e.g. "-" and "__" would result in the same filename. This is probably also one of the reasons the hash above was introduced.